### PR TITLE
Fixed Kudzu Spread Falling Asleep After Containment

### PR DESCRIPTION
**Pull request by Bixkitts**
## About the PR
Modifies a check that fixes the kudzu falling asleep when contained as
described in Issue #15493.

Originally, kudzu entities would stop being an active spreader as soon as they had
no more space to spread.
Now they stop being an active spreader when they are surrounded by other kudzu blocks.

This will impact performance, but (what I hope is) only the necessary amount to make kudzu work properly
with it's current spreading system.
Performance hit could maybe be offset by dynamically changing the SpreadCooldown?
I also commented the spreader system (perhaps too) verbosely as I worked, can roll that back if needed.

Alternative solution would perhaps be to subscribe to damage events on neighboring non-kudzu entities.

**Media**

- this PR does not require an ingame showcase

**Changelog**

:cl:
- fix: Kudzu spread falling asleep when blocked by walls

### CODE REVIEW

File: Content.Server/Spreader/KudzuSystem.cs [hunk 1]
```diff
@@ -52,7 +52,7 @@ private void OnKudzuSpread(EntityUid uid, KudzuComponent component, ref SpreadNe
             return;
         }
 
-        if (args.NeighborFreeTiles.Count == 0)
+        if (args.Neighbors.Count == 4)
         {
             RemCompDeferred<EdgeSpreaderComponent>(uid);
             return;
```
Comment (metalgearsloth) [inline Content.Server/Spreader/KudzuSystem.cs:58]: Shouldn't this be getting added on in the airtight changed sub on spreadersystem? Might be better looking at that.

Comment (Bixkitts) [reply Content.Server/Spreader/KudzuSystem.cs:58]: Looking into it right now.
I haven't got all up in atmos and nodegroup stuff yet but AirtightComponent seems indicate if an entity blocks
atmos or not, and in what direction.

The OnAirtightChanged Event only appears to be raised by AirtightRotate(), and SetAirblocked().
Incidentally, SetAirblocked() is called by Door entities (ONLY door entities) when toggling which is the only thing I quickly found that should raise the OnAirtightChanged event on spreaders, except it doesn't seem to work. 
It should add an EdgeSpreader component to kudzu neighboring the toggling door except it doesn't. 
I haven't seen OnAirtightChanged get raised by walls at all. 

How the kudzu wakes up and keeps growing when attacked still eludes me, as AirtightChanged is the
only thing that seems to add Edgespreader components back and the kudzu isn't Airtight.

I think having the kudzu remove their EdgeSpreaderComponent only when surrounded by more kudzu like in my check is probably more performant than subscribing to all airtight or damage events anyways, but I'm new to the codebase and C#.

Comment (Bixkitts) [reply Content.Server/Spreader/KudzuSystem.cs:58]: OnAirtightChanged also appears to get raised on AirtightComponent ComponentShutdown which I assume
is raised when the component gets removed/deleted i.e. when I would remove a wall from the level.
Something about that just isn't working right.

**Comment by Bixkitts** [comment]
A toggling door/wall removal ends up successfully calling SetAirblocked() in the AirtightSystem, which should raise the OnAirtightChanged event in the SpreaderSystem but it just doesn't and so the spreader doesn't react.
image
It's a server based system raising a server based local event, what am I missing here?

**Comment by Bixkitts** [comment]
Turns out, it wasn't set to broadcast.
Now it's being called, but that still doesn't fix the issue.
Something is going wrong in the SpreaderSystem.GetNeighbors() function.

**Comment by Bixkitts** [comment]
The TransformComponent of the wall being destroyed is lost when it's destroyed, so the SpreaderSystem can't get the coordinates to
check for spreaders it needs to wake up when a wall next to it is destroyed.
Kudzu now spreads through doors fine though, when they are opened.

**Comment by Bixkitts** [comment]
Merged as is, kudzu will now spread through opening doors (game changer).
To make it work this way for walls though, I'd need to find a way to pass the wall's coordinates
through the AirtightChanged event when it's destroyed, which at first thought seems unwieldy.

File: Content.Server/Spreader/SpreaderSystem.cs [hunk 2]
```diff
@@ -87,18 +87,20 @@ private void OnGridUnpaused(EntityUid uid, SpreaderGridComponent component, ref
...
     }
 
     /// <inheritdoc/>
     public override void Update(float frameTime)
     {
         var curTime = _timing.CurTime;
 
-        // Check which grids are valid for spreading.
+        // Check which grids are valid for spreading
+        // Every map seems to have a single "SpreaderGrid" component
```
Comment (metalgearsloth) [inline Content.Server/Spreader/SpreaderSystem.cs:98]:
```suggestion
        // Check which grids are valid for spreading
```

Don't need to muse in comments.

File: Content.Server/Spreader/SpreaderSystem.cs [hunk 3]
```diff
@@ -87,18 +87,20 @@ private void OnGridUnpaused(EntityUid uid, SpreaderGridComponent component, ref
...
         var spreadable = new ValueList<EntityUid>();
         var spreadGrids = EntityQueryEnumerator<SpreaderGridComponent>();
 
+        // The following while() and if() essentially just let execution continue
+        // one time every second (SpreadCooldown)
```
Comment (metalgearsloth) [inline Content.Server/Spreader/SpreaderSystem.cs:103]:
```suggestion
```

File: Content.Server/Spreader/SpreaderSystem.cs [hunk 4]
```diff
@@ -137,17 +144,22 @@ public override void Update(float frameTime)
                 continue;
             }
 
+            // _spreaderGroups is a string list of the different
+            // possible spreader types, as of writing "kudzu", "puddle"
+            // or "smoke"
```
Comment (metalgearsloth) [inline Content.Server/Spreader/SpreaderSystem.cs:149]:
```suggestion
```

File: Content.Server/Spreader/SpreaderSystem.cs [hunk 5]
```diff
@@ -319,6 +337,7 @@ public List<EntityUid> GetNeighbors(EntityUid uid, AirtightComponent comp)
         if (!EntityManager.TryGetComponent<TransformComponent>(uid, out var transform))
             return neighbors; // how did we get here?
 
+        // The entity will have been removed by the time it gets here
```
Comment (metalgearsloth) [inline Content.Server/Spreader/SpreaderSystem.cs:340]:
```suggestion
```

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**